bp_trap.c

```diff
@@ -0,0 +1,86 @@
+#include <stdint.h>
+#include "bp_trap.h"
+#include "emulation.h"
+
+#define EXCEPTION_TOTAL            16
+#define INTERRUPT_TOTAL            12
+
+static void (*exception_handler[EXCEPTION_TOTAL])(uint64_t *, uint64_t, uint64_t) =
+{
+  bp_unhandled_trap_abort,        // 0
+  bp_unhandled_trap_abort,        // 1
+  bp_emulate_illegal_instruction, // 2
+  bp_unhandled_trap_abort,        // 3
+  bp_unhandled_trap_abort,        // 4
+  bp_unhandled_trap_abort,        // 5
+  bp_unhandled_trap_abort,        // 6
+  bp_unhandled_trap_abort,        // 7
+  bp_unhandled_trap_abort,        // 8
+  bp_unhandled_trap_abort,        // 9
+  bp_unhandled_trap_abort,        // 10
+  bp_unhandled_trap_abort,        // 11
+  bp_unhandled_trap_abort,        // 12
+  bp_unhandled_trap_abort,        // 13
+  bp_unhandled_trap_abort,        // 14
+  bp_unhandled_trap_abort         // 15
+};
+
+static void (*interrupt_handler[INTERRUPT_TOTAL])(uint64_t *, uint64_t, uint64_t) =
+{
+  bp_unhandled_trap_abort,        // 0
+  bp_unhandled_trap_abort,        // 1
+  bp_unhandled_trap_abort,        // 2
+  bp_unhandled_trap_abort,        // 3
+  bp_unhandled_trap_abort,        // 4
+  bp_unhandled_trap_abort,        // 5
+  bp_unhandled_trap_abort,        // 6
+  bp_unhandled_trap_abort,        // 7
+  bp_unhandled_trap_abort,        // 8
+  bp_unhandled_trap_abort,        // 9
+  bp_unhandled_trap_abort,        // 10
+  bp_unhandled_trap_abort,        // 11
+};
+
+int register_trap_handler(void (*handler)(uint64_t *, uint64_t, uint64_t),
+        int index, int is_exception)
```

Comment:
You should directly pass cause, is exception vs is trap can be detected by comparing to 0 (interrupts are negative)